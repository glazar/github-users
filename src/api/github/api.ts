import { GithubUser } from './api';
export interface GithubUser {
  id: number;
  login: string;
  avatar_url: string;
  html_url: string;
}

export const fetchUserList = (since?: number): Promise<GithubUser[]> => {
  let url = 'https://api.github.com/users';

  if (since) {
    url += `?since=${since}`;
  }

  return fetch(url).then(response => response.json<GithubUser[]>());
};

export const fetchUser = (login: string): Promise<GithubUser> => {
  const url = `https://api.github.com/users/${login}`;

  return fetch(url).then(response => {
    if (response.status === 404) {
      throw response;
    } else {
      return response.json<GithubUser>();
    }
  });
};
