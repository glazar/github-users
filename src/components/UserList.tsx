import { PureComponent } from 'react';
import { SFC } from 'react';
import * as React from 'react';
import { Alert, Button, ListGroup, ListGroupItem, PageHeader } from 'react-bootstrap';
import { connect } from 'react-redux';
import { LinkContainer } from 'react-router-bootstrap';
import { ThunkDispatch } from 'redux-thunk';

import { getNextUserList, getUserList, GetUserListActions } from '../actions/userListActions';
import { GithubUser } from '../api/github/api';
import { State, UserListState } from '../store';
import { Loader } from './Spinner';
import { UserThumbnail } from './UserThumbnail';

interface UserListDispatchProps {
  onReloadUserList?: () => void;
  onLoadUserList?: () => void;
}

const mapStateToProps = (state: State): UserListState => {
  return {
    userList: state.userListState.userList,
    isFetching: state.userListState.isFetching,
    canFetchMore: state.userListState.canFetchMore
  };
};

const mapDispatchToProps = (dispatch: ThunkDispatch<State, any, GetUserListActions>): UserListDispatchProps => {
  return {
    onReloadUserList() {
      dispatch(getUserList());
    },
    onLoadUserList() {
      dispatch(getNextUserList());
    }
  };
};

type UserListProps = UserListState & UserListDispatchProps;

export class UserList extends PureComponent<UserListProps, {}> {
  public componentDidMount(): void {
    if (!this.props.userList) {
      this.handleReloadUserList();
    }
  }

  public render(): JSX.Element {
    const { userList, isFetching, canFetchMore } = this.props;

    return (
      <div>
        <PageHeader>User List</PageHeader>
        {!userList && !isFetching && <Alert bsStyle="warning">No Data Available</Alert>}

        {userList && (
          <>
            {!isFetching && (
              <Button bsStyle="primary" className="m-2" bsSize="large" onClick={this.handleReloadUserList}>
                Reload Users
              </Button>
            )}
            <UserCardList userList={userList} />
          </>
        )}
        {isFetching && <Loader />}
        {canFetchMore &&
          !isFetching && (
            <Button bsStyle="primary" className="m-2" bsSize="large" onClick={this.handleLoadUserList}>
              Load More Users
            </Button>
          )}
      </div>
    );
  }

  private handleReloadUserList = (): void => {
    this.props.onReloadUserList!();
  };

  private handleLoadUserList = (): void => {
    this.props.onLoadUserList!();
  };
}

export const UserCardList: SFC<{ userList: GithubUser[] }> = ({ userList }) => (
  <ListGroup className="border-0">
    {userList.map(user => (
      <ListGroupItem key={user.id} className="border-0">
        <UserCard user={user} />
      </ListGroupItem>
    ))}
  </ListGroup>
);

const UserCard: SFC<{ user: GithubUser }> = ({ user }) => (
  <UserThumbnail user={user}>
    <p>
      <LinkContainer to={`/user/${user.login}`}>
        <Button bsStyle="info" bsSize="large">
          Details
        </Button>
      </LinkContainer>
    </p>
  </UserThumbnail>
);

export const ConnectedUserList = connect(
  mapStateToProps,
  mapDispatchToProps
)(UserList);

export default ConnectedUserList;
