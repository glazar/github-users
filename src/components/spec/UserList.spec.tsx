import { shallow } from 'enzyme';
import * as React from 'react';

import { GithubUser } from '../../api/github/api';
import { UserCardList, UserList } from '../UserList';

it('UserCardList renders user data properly', () => {
  const userList: GithubUser[] = [
    {
      id: 1,
      avatar_url: 'https://avatars3.githubusercontent.com/u/4364303?v=4',
      login: 'glazar',
      html_url: 'https://github.com/glazar'
    }
  ];
  const wrapper = shallow(<UserCardList userList={userList} />);
  expect(wrapper).toMatchSnapshot();
});

it('UserList component should call to reload user list if no user list provided', () => {
  const mockHandleReloadUserList = jest.fn();

  shallow(<UserList onReloadUserList={mockHandleReloadUserList} />);
  expect(mockHandleReloadUserList).toHaveBeenCalledTimes(1);
});

it('UserList component should not call to reload user list if user list provided', () => {
  const mockHandleReloadUserList = jest.fn();

  const userList: GithubUser[] = [
    {
      id: 1,
      avatar_url: 'https://avatars3.githubusercontent.com/u/4364303?v=4',
      login: 'glazar',
      html_url: 'https://github.com/glazar'
    }
  ];

  shallow(<UserList userList={userList} onReloadUserList={mockHandleReloadUserList} />);
  expect(mockHandleReloadUserList).not.toHaveBeenCalled();
});
