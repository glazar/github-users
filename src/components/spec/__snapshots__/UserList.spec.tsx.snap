// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCardList renders user data properly 1`] = `
<ListGroup
  bsClass="list-group"
  className="border-0"
>
  <ListGroupItem
    bsClass="list-group-item"
    className="border-0"
    key="1"
    listItem={false}
  >
    <UserCard
      user={
        Object {
          "avatar_url": "https://avatars3.githubusercontent.com/u/4364303?v=4",
          "html_url": "https://github.com/glazar",
          "id": 1,
          "login": "glazar",
        }
      }
    />
  </ListGroupItem>
</ListGroup>
`;
